### DISCUSSION TITLE
Increase `innodb_page_size` for mysql

### DISCUSSION THREAD

**Discussion by leevigraham** [ddev/ddev#6543] [Q&A] [answered]
Is there a way to configure the `innodb_page_size` setting in mysql so that it applies **before** the mysql container is created and the db is created?

I tried to add a custom `mysql.cnf` but I received an error:

The error  is caused by a mismatch between the page size in your existing InnoDB data files and the innodb_page_size value that you're trying to start MySQL with. Specifically, your data file (./ibdata1) was created with a page size of 16KB (16384), but you're now trying to start MySQL with a page size of 64KB (65536). InnoDB cannot change the page size for existing data files.

When importing I get the following error:

```
 error: The shell command "mysql --defaults-extra-file="/tmp/vxwivaqumtvg.cnf" db < "./tests/db.sql"" failed with exit code 1: ERROR 1118 (42000) at line 1723: Row size too large. The maximum row size for the used table type, not counting BLOBs, is 8126. This includes storage overhead, check the manual. You have to change some columns to TEXT or BLOBs
```

by default the innodb_page_size is 16k

```sh
mysql> SHOW VARIABLES LIKE "innodb_page_size";
+------------------+-------+
| Variable_name    | Value |
+------------------+-------+
| innodb_page_size | 16384 |
+------------------+-------+
1 row in set (0.01 sec)
```

**Comment by rfay** [comment]
Can you say why you want to change `innodb_page_size` ? 
* I imagine you could in fact fundamentally override it with a `.ddev/db_build/Dockerfile.innodb` that changed the /etc/my.cnf. 
* You might be able to do it with your current strategy when creating a new database, perhaps with `ddev import-db --database`

The problem is that the `db` database is precreated and bundled into the image, so that's why you're not getting access to change it.

But I doubt that changing this value is a good idea. Interested in why you think it is.

**Reply by leevigraham** [reply]
In my case it's because of a CraftCMS plugin (freeform) adding 273 columns (some TEXT some varchar). To many fields throwing an error on import.

**Reply by rfay** [reply]
I haven't heard of making the page size larger as a solution to too many columns. Do you have a link for that?

**Reply by leevigraham** [reply]
I'm not an expert so this will need some more documentation.

https://dev.mysql.com/doc/refman/8.4/en/column-count-limit.html

> Row Size Limits: The maximum row size for an InnoDB table, which applies to data stored locally within a database page, is slightly less than half a page for 4KB, 8KB, 16KB, and 32KB [innodb_page_size](https://dev.mysql.com/doc/refman/8.4/en/innodb-parameters.html#sysvar_innodb_page_size) settings. For example, the maximum row size is slightly less than 8KB for the default 16KB InnoDB page size. For 64KB pages, the maximum row size is slightly less than 16KB. See [Section 17.21, “InnoDB Limits”](https://dev.mysql.com/doc/refman/8.4/en/innodb-limits.html).

**Reply by rfay** [reply]
I'm sure you know already that this *very* significant design flaw in Craft CMS (Using an extra column for every single thing) was fixed in Craft 5. 

I guess DDEV doesn't specify innodb_page_size so it takes the default 16KB, https://dev.mysql.com/doc/refman/8.0/en/innodb-init-startup-configuration.html

I'm probably open to increasing innodb_page_size, if we could figure out how to do that without breaking existing projects.

**Reply by rfay** [reply]
I haven't found a solution to this, but ChatGPT suggested `ALTER TABLE tablename ROW_FORMAT=COMPRESSED;` as an option for this problem. I'd be interested if that affects anything.

Since this is a longstanding (and probably well-understood Craft CMS problem) I assume you've already conferred with the Craft community about it?

**Reply by leevigraham** [reply]
Craft 5 rewrote their content storage to a large json blob to tackle this issue. There doesn't seem to be solution for v3 / v4

**Reply by rfay** [reply]
Being an old-school relational database normalization person, I still can't quite approve of the json blob approach :(

**Reply by rfay** [reply]
Could you please try the artifacts in https://github.com/ddev/ddev/pull/6549#issuecomment-2356021696 and see if they give you any hope? You'll need to `ddev delete` or try out in a fresh project. 

It may be that the way to do this is by adding a new setting for a DDEV project where it doesn't create the database by default during start on a new project. I'm not sure what the impact of that would be. The reason DDEV precreates the db (at image build time) is that it speeds up startup quite a lot.

**Reply by leevigraham** [reply]
I'll take a look.

**Reply by leevigraham** [reply]
@rfay Sorry for the delayed response. I can confirm the artifact in [#6549 ](https://github.com/ddev/ddev/pull/6549#issuecomment-2356021696) worked importing the DB with no other changes.

I'm going to double check now by reverting back to the current version of DDEV and trying to import again. Then I'll try tweak the DB table to use `ROW_FORMAT=dynamic` as a SQL fix.

**Reply by leevigraham** [reply]
Successful import:

<img width="519" alt="image" src="https://github.com/user-attachments/assets/03a478ae-d94d-4286-9628-4b0e5c42d001">

<img width="1638" alt="image" src="https://github.com/user-attachments/assets/86193aff-a41e-42ec-9b1c-90a18f3bdca0">

Failed import:

<img width="557" alt="image" src="https://github.com/user-attachments/assets/848a0741-47cd-490c-b6bf-a795b3d92f45">

<img width="1642" alt="image" src="https://github.com/user-attachments/assets/eb325091-901c-4596-9837-d6c880bf78d3">

**Reply by rfay** [reply]
So that fixes it for you, right? Did you have to do anything else?

**Reply by leevigraham** [reply]
Only thing I changed in the successful test was using the artifact ddev with `ddev start`. 

Obviously this is still a DB schema issue with the number of columns of a third party Craft plugin. I'm not 100% sure DDEV should be solving this particular use case *but* having a way to change some of the mysql config (eg `innodb_page_size`) would be useful.

**Reply by rfay** [reply]
Thank you. I'm trying to reach out to some database experts to see how much risk there is in this change and whether it might impact anything badly. If the answer is positive, maybe it can go into the next major release. 

I'd appreciate it if you'd reach out to any Craft CMS folks you know and ask for a review. I imagine Pixel & Tonic is all on board with Craft 5, so maybe won't have any interest. I do wonder how many people run into this though.

**Reply by leevigraham** [reply]
Related https://github.com/craftcms/cms/issues/7750 https://github.com/craftcms/cms/issues/2009

**Reply by leevigraham** [reply]
Raised a discussion here with the plugin devs: https://github.com/solspace/craft-freeform/discussions/1544

**Reply by rfay** [reply]
Thanks for the work chasing this.  @brandonkelly 's argument on the PR was certainly valid, if you can't solve this on local you're going to have a hard time solving it for a hosting environment.

**Reply by leevigraham** [reply, marked as answer]
@rfay After some more investigation with the client (who manages their own mysql docker container) I found that they have set `innodb_strict_mode=OFF` in production.

**Reply by rfay** [reply]
Are you able to add that in your .ddev/mysql/<whatever>.cnf and make it work for you?

**Reply by leevigraham** [reply]
Yep in this case it resolves the row size "error". I still think it would be nice to have a way to update `innodb_page_size` as a config in ddev… although it's not an immediate need for me atm.

**Reply by rfay** [reply]
It would, but that would make every `ddev start` significantly slower for every user of DDEV, like at least 10-20 seconds slower, while it did the things that are normally prebuilt in the database.
